blocks/monitoring/evaluators.py

```diff
@@ -211,8 +248,23 @@ def process_batch(self, batch):
         if self._accumulate_fun is not None:
             self._accumulate_fun(**batch)
 
+        try:
+            batch = dict_subset(batch, self.buffer2_.input_names)
+        except KeyError:
+            reraise_as(
+                "Not all data sources required for monitoring were"
+                " provided. The list of required data sources:"
+                " {}.".format(self.buffer2_.input_names))
+        if self._accumulate_fun2 is not None:
```

Comment:
I would move this logic to the `NonTheanoBuffer`. That is `DatasetEvaluator` calls the functions, gets the numerical values and passes it the `NonTheanoBuffer` that knows better what to do with them.